node/pegnet/txhistory.go

```diff
@@ -0,0 +1,448 @@
+package pegnet
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/Factom-Asset-Tokens/factom"
+	"github.com/pegnet/pegnet/modules/grader"
+	"github.com/pegnet/pegnetd/fat/fat2"
+)
+
+// HistoryAction are the different types of actions inside the history
+type HistoryAction int32
+
+const (
+	// Invalid is used for debugging
+	Invalid HistoryAction = iota
+	// Transfer is a 1:n transfer of pegged assets from one address to another
+	Transfer
+	// Conversion is a conversion of pegged assets
+	Conversion
+	// Coinbase is a miner reward payout
+	Coinbase
+	// FCTBurn is a pFCT payout for burning FCT on factom
+	FCTBurn
+)
+
+// HistoryTransaction is a flattened entry of the history table structure.
+// It contains several actions: transfers, conversions, coinbases, and fct burns
+type HistoryTransaction struct {
+	Hash      *factom.Bytes32 `json:"hash"`
+	Height    int64           `json:"height"`
+	Timestamp time.Time       `json:"timestamp"`
+	Executed  int32           `json:"executed"`
+	TxIndex   int             `json:"txindex"`
+	TxAction  HistoryAction   `json:"txaction"`
+
+	FromAddress *factom.FAAddress          `json:"fromaddress"`
+	FromAsset   string                     `json:"fromasset"`
+	FromAmount  int64                      `json:"fromamount"`
+	ToAsset     string                     `json:"toasset,omitempty"`
+	ToAmount    int64                      `json:"toamount,omitempty"`
+	Outputs     []HistoryTransactionOutput `json:"outputs,omitempty"`
+}
+
+// HistoryTransactionOutput is an entry of a transfer's outputs
+type HistoryTransactionOutput struct {
+	Address factom.FAAddress `json:"address"`
+	Amount  int64            `json:"amount"`
+}
+
+// in the context of tables, `history_txbatch` is the table that holds the unique reference hash
+// and `transaction` is the table that holds the actions associated with that unique reference hash
+// `lookup` is an outside reference that indexes the addresses involved in the actions
+//
+// associations are:
+// 	* history_txbach : history_transaction is `1:n`
+// 	* history_transaction : lookup is `1:n`
+//	* lookup : (transaction.outputs + transaction.inputs) is `1:n` (unique addresses only)
+const createTableTxHistoryBatch = `CREATE TABLE IF NOT EXISTS "pn_history_txbatch" (
+	"history_id"	INTEGER PRIMARY KEY,
+	"entry_hash"    BLOB NOT NULL,
+	"height"        INTEGER NOT NULL, -- height the tx is in
+	"blockorder"	INTEGER NOT NULL,
+	"timestamp"		INTEGER NOT NULL,
+	"executed"		INTEGER NOT NULL, -- -1 if failed, 0 if pending, height it was applied at otherwise
+
+	UNIQUE("entry_hash", "height"),
+	FOREIGN KEY("height") REFERENCES "pn_grade"
```

Comment:
Is there a reason why we reference pn_grade as a foreign key?

I don't see why we'd need to query the pn_grade table when asking for txs?

Reply:
ah yeah, that's leftover from the way i did it before (i pulled the rate from pn_grade in the query)